### ISSUE TITLE
Breaking change for Prophet Weekly Seasonality

### ISSUE THREAD

**Issue by nabeel-oz** [nabeel-oz/qlik-py-tools#77]
**Function**
Prophet_Seasonality

**Describe the bug**
Release 6.0 included changes to the usage of the `Prophet_Seasonality` function when calculating Weekly seasonality (commit 72034a9024c4dc56666493f0b13266b19bc412c5). Expressions using this capability will stop working until the fix described below is applied.

**Fix**
The weekly seasonality now requires the Weekday function to be used in the dimension and the measure. This change allows the result to always be returned in the correct order.

The documentation and sample app were updated to reflect the changes: https://github.com/nabeel-oz/qlik-py-tools/blob/master/docs/Prophet.md#seasonality

In short, you'll need to update the dimension for your chart to:

```
Weekday(FORECAST_DATE)
```

And the measure will be:

```
PyTools.Prophet_Seasonality(Max(Weekday(FORECAST_DATE)), $(vAccidentsByDate), '', 'freq=D, return=weekly, seasonality=weekly')
```

**Comment by Odassier** [comment]
Hi Nabeel,

First, thank you for this extension, it is amazing, really look forward to using it more. I tried to replicate the simple forecasting example but am encountering an error message on the weekly seasonality.  
"ValueError: 2 columns passed, passed data had 1 columns" but my variable works, I get the string of values when I open it in a KPI object.  Not sure why this is happening. Can you have a look and let me know what I am doing wrong?

Dimension: =Weekday([Prophet.Date])
Measure: PyTools.Prophet_Seasonality(Max(Weekday([Prophet.Date])), $(vROOFbyDATE), '', 'freq=D, return=weekly, seasonality=weekly')

Here is the full debug script if that will help.
2019-12-15 14:11:48,441 - INFO - ipv6:[::1]:59335 - Capability 'Prophet_Seasonality' called by user Personal\Me from app C:\Users\odassier\Documents\Qlik\Sense\Apps\Axioma ROOF Forecasting.qvf
2019-12-15 14:11:48,441 - INFO - ExecuteFunction (functionId: 8, _prophet_seasonality)
2019-12-15 14:11:48,443 - ERROR - Exception iterating responses: 2 columns passed, passed data had 1 columns
Traceback (most recent call last):
  File "C:\Users\odassier\Documents\Qlik\Sense\Extensions\qlik-py-tools-6.2\qlik-py-env\lib\site-packages\pandas\core\internals\construction.py", line 500, in _list_to_arrays
    content, columns, dtype=dtype, coerce_float=coerce_float
  File "C:\Users\odassier\Documents\Qlik\Sense\Extensions\qlik-py-tools-6.2\qlik-py-env\lib\site-packages\pandas\core\internals\construction.py", line 583, in _convert_object_array
    "{con} columns".format(col=len(columns), con=len(content))
AssertionError: 2 columns passed, passed data had 1 columns

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\odassier\Documents\Qlik\Sense\Extensions\qlik-py-tools-6.2\qlik-py-env\lib\site-packages\grpc\_server.py", line 453, in _take_response_from_response_iterator
    return next(response_iterator), True
  File "__main__.py", line 420, in _prophet_seasonality
    predictor = ProphetForQlik.init_seasonality(request_list, context)
  File "C:\Users\odassier\Documents\Qlik\Sense\Extensions\qlik-py-tools-6.2\qlik-py-env\core\_prophet.py", line 159, in init_seasonality
    request_df = pd.DataFrame([p.split(":") for p in pairs], columns=['ds', 'y'])
  File "C:\Users\odassier\Documents\Qlik\Sense\Extensions\qlik-py-tools-6.2\qlik-py-env\lib\site-packages\pandas\core\frame.py", line 450, in __init__
    arrays, columns = to_arrays(data, columns, dtype=dtype)
  File "C:\Users\odassier\Documents\Qlik\Sense\Extensions\qlik-py-tools-6.2\qlik-py-env\lib\site-packages\pandas\core\internals\construction.py", line 464, in to_arrays
    return _list_to_arrays(data, columns, coerce_float=coerce_float, dtype=dtype)
  File "C:\Users\odassier\Documents\Qlik\Sense\Extensions\qlik-py-tools-6.2\qlik-py-env\lib\site-packages\pandas\core\internals\construction.py", line 503, in _list_to_arrays
    raise ValueError(e) from e
ValueError: 2 columns passed, passed data had 1 columns

In advance, thank you for your help on this.

Olivier

**Comment by nabeel-oz** [comment]
Hi Olivier,

The error message indicates that you have a value in your input series which can't be converted to the 'ds' and 'y' columns correctly. This may be a null value, or a pair where ds or y are missing.

To investigate you can break the `vROOFbyDATE` down into it's component dimension and measure and view the output in a table like I've done below. 

![image](https://user-images.githubusercontent.com/21298843/70871294-f10d9500-1ff1-11ea-8cad-6bc5c8f24c32.png)

If that doesn't help, I'd check if there are any null rows when using `Weekday([Prophet.Date])` dimension with `vROOFbyDATE`.

**Comment by Odassier** [comment]
Hi Nabeel,

Thanks for this.  I checked (see screenshot below) but there are no Null
values (although in the data file I have "N/A" for the five extra dates I
want a forecast for).

[image: image.png]

I'm attaching the ,qvf file here if you can check.

thanks,

Olivier

**Comment by Odassier** [comment]
I replaced the "N/A" in the data file with "0" for the five forecast dates,
but it still gives me the same error message even though the table shows no
more "N/A" so all the values are now numbers.

Olivier

**Comment by Odassier** [comment]
Any updated help on this issue? Ciuld you open my app? 

Thanks 

Olivier 

⁣Get BlueMail for Android ​

**Comment by nabeel-oz** [comment]
Hi Olivier, 

The app and screenshot aren't showing up in your post. You can try attaching the QVF as a ZIP file.

Having NULL values for the future dates is fine. The problem seems to be that at least one of the items in your variable can't be split into the two columns 'ds' and 'y' using the ':' separator. 

For example this is incorrect as the second item can't be broken up into two columns:
```
41275:24; 41276; 41277:30; ...
```
This is also incorrect because of the trailing semicolon:
```
41275:24; 41276:27; 41277:30; ...; 
```

Cheers,
Nabeel

**Comment by Odassier** [comment]
[Axioma ROOF Forecasting.zip](https://github.com/nabeel-oz/qlik-py-tools/files/3976174/Axioma.ROOF.Forecasting.zip)

Here you go. the error is in the second sheet. I'm also trying to setup the multivariate test in fubsequent sheets.

Olivier

**Comment by Odassier** [comment]
ok, I see now that my vROOFbyDate variable only has one vector as per your message since there are no colons anywhere. Not sure what I did wrong in the expression for the variable though.
(1823.2923611111;1823.3340277778;1823.3756944444;1823.4173611111;1823.4590277778;1823.5840277778;1823.6256944444;1823.6673611111;1823.7090277778;1823.7506944444;1823.8756944444;1823.9173611111;1823.9590277778;1824.0006944444;1824.0423611111;1824.1673611111;1824.2090277778;1824.2506944444;1824.2923611111;1824.3340277778;1824.4590277778;1824.5006944444;1824.5423611111;1824.5840277778;1824.6256944444;1824.6673611111;1824.7090277778;1824.7506944444;1824.7923611111;1824.8340277778;1824.9173611111;1825.0423611111;1825.0840277778;1825.1256944444;1825.1673611111;1825.2090277778;1825.3340277778;1825.3756944444;1825.4173611111;1825.4590277778;1825.5006944444
--

**Comment by nabeel-oz** [comment]
Your variable is not giving the correct output:

![image](https://user-images.githubusercontent.com/21298843/71048067-1f29db00-2192-11ea-8b1e-040ee7a18184.png)

The variable is meant to provide the full timeseries in the format `ds1:y1;ds2:y2;...`. The output above looks like garbage.

The problem is missing parenthesis in the measure of your `Aggr` expression in `vROOFbyDATE`. The expression below will first execute the `&` operation, the output being a Dual as is the norm in Qlik, meaning it has a text as well as a numeric representation. This will then be used in the `-` operation. In short the `-` operator doesn't take precedence as you expected.

```
Num([Prophet.Date]) & ':' & count([Prophet.Data field]) - Count({$<[Prophet.Data field] = {"N/A"}>} Distinct [Prophet.Data field])
```

You need parenthesis around the subtraction of the two count operations. Updating the expression as below fixes the output:

```
Concat(DISTINCT TOTAL Aggr(Num([Prophet.Date]) & ':' & (count([Prophet.Data field]) - Count({$<[Prophet.Data field] = {"N/A"}>} Distinct [Prophet.Data field])), [Prophet.Date]), ';')
```

![image](https://user-images.githubusercontent.com/21298843/71048476-7bd9c580-2193-11ea-950e-4fe1c707e753.png)

**Comment by Odassier** [comment]
You're right, it works, I understand the logic now.  Just one thing, it works when I select US-LMS_ROOF or JP-LMS_ROOF in the filter box in the first sheet but not for any of the other markets (says the selection returned no data for this chart). Not sure why.

Olivier

**Comment by Odassier** [comment]
Both Japan and US have two and one day with no data in the history, so their count(data field) that is not "N/A" is shorter than the other markets, is that why?  how do you fix that?

**Comment by nabeel-oz** [comment]
If you look at the SSE log you should see that in certain cases you are not sending any actual values at all. Note how y values are all NaN and as a result the Input DataFrame is empty.

![image](https://user-images.githubusercontent.com/21298843/71056093-134b1280-21ac-11ea-9b57-e0c31066bf63.png)

I see that for some charts you don't have an aggregation function for the measure being passed to `PyTools.Prophet`. That's probably the issue.

**Comment by Odassier** [comment]
Had a look and am not sure where any aggregation function should be.  The only chart that isn't working is the days of the week seasonality.

the dimension is: Weekday([Prophet.Date])

The measure is: PyTools.Prophet_Seasonality(Max(Weekday([Prophet.Date])), $(vROOFbyDATE), '', 'freq=D, return=weekly, seasonality=weekly')

and the variable is what we just fixed: Concat(DISTINCT TOTAL Aggr(Num([Prophet.Date]) & ':' & (Count([Prophet.Data field]) - Count({$<[Prophet.Data field] = {"N/A"}>} Distinct [Prophet.Data field])), [Prophet.Date]), ';')

All I do is change the selection in the filter called "Market" which is based on "Propher.Data.Attribute".  Now I do not see that reference in the dimension or measure of the chart but since this is an unpivoted table in Qlik, the Prophet.Data.Attribute field which was created automatically when uunpivoting should represent a dimension I can filter the data on.

Not sure why it sends the data for Japan (JP-LMS) and US (IS-LMS) but not the rest of the Prophet.Data Attribute fields.

Olivier

**Comment by nabeel-oz** [comment]
I was referring to the trend and weekly seasonality charts on your first sheet. The `Prophet.Data field` is wrapped up with a `Sum` in the top line chart but not in the bottom two which causes issues without selections or when multiple values are selected. Also, if `Sum` is the correct aggregation for `Prophet.Data field`, shouldn't you be using it in your `vROOFbyDATE` expression as well instead of `Count`?

I didn't test the second sheet for the selections you mention, but you should just approach this as a Qlik data modelling problem. To figure out where things are going wrong it helps to break down your expressions into smaller components, add them to a table and test the output with those selections.

**Comment by Odassier** [comment]
Good catch. Fixed it, thanks. 
Since I can fiddle with some of the additional arguments (trend fitting, change points, etc.), is there a 'goodness-of-fit' I can call via PyTools to let me know when I've got the optimal settings for this time series? How do I know when I've got the best forecast I can get?
thanks,
Olivier

**Comment by nabeel-oz** [comment]
There is no fixed way of telling whether you've got the best forecast possible with Prophet. It's usually more a case of experimentation to find a 'good enough' model. 

PyTools doesn't provide evaluation metrics for Prophet, but you can calculate these in Qlik. For example add the Predicted and Actual values to a table and use the expression `Column(1) - Column(2)` to get the absolute error.

**Comment by Odassier** [comment]
Yes, that's what I did.  Thanks for letting me know there is no other way for now.  You can close this issue.
Happy Holidays,

Olivier
